CHANGELOG.md

```diff
@@ -3,6 +3,360 @@
+| `CreateApplicationCommandPermissionData` | `CreateCommandPermission` |
+| `CreateApplicationCommandPermissionsData` | `EditCommandPermissions` |
+| `CommandPermission` | `CommandPermissions` |
+| `CommandPermissionData` | `CommandPermission` |
+
+Also, the methods on `CreateCommandPermission`, such as `new`, `kind`, etc. have been replaced with constructors for each type of permission, e.g. `role`, `user`, `channel`, etc., to avoid a possible mismatch between `kind` and the id that gets passed in.
+
+Finally, the `{GuildId,Guild,PartialGuild}::create_command_permission` method has been renamed to `edit_command_permission`.
+
+### Cache
+
+* Cache methods now (mostly) return a `CacheRef` type that wraps a reference into the internal `DashMap` used by the cache. Some others now also return a wrapper type around the whole `DashMap`, but with a limited API to prevent accidental deadlocks. This all helps reduce the number of clones when querying the cache. Those wishing to replicate the old behavior can simply clone the references wholesale to obtain the wrapped data.
+* `CacheSettings` has new fields `time_to_live`, `cache_guilds`, `cache_channels`, and `cache_users`, allowing cache configuration on systems with memory requirements.
+
+### Interactions
+
+The various interaction types have been renamed as follows:
+
+| serenity v0.11 | serenity v0.12 |
+| --- | --- |
+| `ApplicationCommandInteraction` | `CommandInteraction` |
+| `MessageComponentInteraction` | `ComponentInteraction` |
+| `ModalSubmitInteraction` | `ModalInteraction` |
+
+Method names on interaction types have been shortened in the following way:
+
+| serenity v0.11 | serenity v0.12 |
+| --- | --- |
+| `create_interaction_response` | `create_response` |
+| `create_followup_message` | `create_followup` |
+| `delete_original_interaction_response` | `delete_response` |
+| `delete_followup_message` | `delete_followup` |
+| `edit_original_interaction_response` | `edit_response` |
+| `edit_followup_message` | `edit_followup` |
+| `get_interaction_response` | `get_response` |
+| `get_followup_message` | `get_followup` |
+
+* `AutocompleteInteraction` has been merged into `CommandInteraction`, along with its corresponding methods.
+* Add `as_*` and `into_*` helper methods to the `Interaction` type for converting to each of its respective variants.
+* The `kind` field has been removed from each of the interaction structs.
+* A `quick_modal` method has been added to `CommandInteraction` and `ComponentInteraction`. See the docs for more details.
+* A `defer_ephemeral` helper method has been added to many interaction types.
+
+### Framework
+
+* The standard framework is now configurable at runtime, as the `configure` method now takes `self` by reference.
+* The `Framework` trait has been reworked - the `dispatch` method now takes a `FullEvent` enum as argument, which allows for more flexibility. Also, there is now an added optional `init` method.
```

Comment:
We could add the motivation for this change:

The standard framework is now configurable at runtime, as the `configure` method now takes `self` by reference.

```md
The `Framework` trait has been reworked to work for much more use cases than just text commands:
* The `dispatch` method now takes a `FullEvent` instead of just `Message`
  * `FullEvent` contains all the data that is passed to `EventHandler`.
* The `Framework` trait gained the `init(&mut self, &Client)` method to execute HTTP requests and access cache and shard data on startup.

With these changes, the `Framework` trait becomes suitable for alternative frameworks like [poise](https://github.com/serenity-rs/poise/).
```

Reply:
Addressed in most recent commit.